python/mcap-ros2-support/mcap_ros2/dynamic.py

```diff
@@ -0,0 +1,354 @@
+    reader = CdrReader(data)
+    return read_complex_type(msgdef.msg_name, msgdef.fields, msgdefs, reader)
+
+
+def read_complex_type(
+    msg_name: str,
+    fields: list[Field],
+    msgdefs: Dict[str, MessageSpecification],
+    reader: CdrReader,
+) -> Message:
+    Msg = type(
+        msg_name,
+        (SimpleNamespace,),
+        {
+            "__name__": msg_name,
+            "__slots__": [field.name for field in fields],
+            "__repr__": __repr__,
+            "__str__": __repr__,
+            "__getstate__": __getstate__,
+            "__setstate__": __setstate__,
+        },
+    )
+    msg = Msg()
+
+    for field in fields:
+        if not field.type.is_primitive_type():
+            # Complex type
+            nested_definition = msgdefs[f"{field.type.pkg_name}/{field.type.type}"]
+            if nested_definition is None:
+                raise ValueError(
+                    f'Message definition not found for field "{field.name}" with '
+                    'type "{field.type.type}"'
+                )
+
+            if field.type.is_array:
+                # For dynamic length arrays we need to read a uint32 prefix
+                array_length = field.type.array_size or reader.uint32()
+                array = [
+                    read_complex_type(
+                        nested_definition.msg_name,
+                        nested_definition.fields,
+                        msgdefs,
+                        reader,
+                    )
+                    for _ in range(array_length)
+                ]
+                setattr(msg, field.name, array)
+            else:
+                value = read_complex_type(
+                    nested_definition.msg_name,
+                    nested_definition.fields,
+                    msgdefs,
+                    reader,
+                )
+                setattr(msg, field.name, value)
+        else:
+            # Primitive type
+            if field.type.is_array:
+                parser_fn = ARRAY_PARSERS[field.type.type]
+                if parser_fn is None:
+                    raise NotImplementedError(
+                        f"Parsing for type {field.type.type}[] is not implemented"
+                    )
+                # For dynamic length arrays we need to read a uint32 prefix
+                array_length = field.type.array_size or reader.sequence_length()
+                value = parser_fn(reader, array_length)
+                setattr(msg, field.name, value)
+            else:
+                parser_fn = FIELD_PARSERS[field.type.type]
+                if parser_fn is None:
+                    raise NotImplementedError(
+                        f"Parsing for type {field.type.type} is not implemented"
+                    )
+                value = parser_fn(reader)
+                setattr(msg, field.name, value)
+
+    return msg
+
+
+def generate_dynamic(schema_name: str, schema_text: str) -> Dict[str, DecoderFunction]:
```

Comment:
this function name could use more info. are we generating a dynamic? maybe generate_msg_class? generate_decoder?